src/analysis/refactor_extract_region.ml

```diff
@@ -0,0 +1,395 @@
+  let subst_loc =
+    let start_lnum =
+      1 + expr.exp_loc.Location.loc_start.pos_lnum
+      - toplevel_item.loc.loc_start.pos_lnum
+    in
+    let end_lnum =
+      start_lnum + expr.exp_loc.loc_end.pos_lnum
+      - expr.exp_loc.loc_start.pos_lnum
+    in
+    { expr.exp_loc with
+      loc_start = { expr.exp_loc.loc_start with pos_lnum = start_lnum };
+      loc_end = { expr.exp_loc.loc_end with pos_lnum = end_lnum }
+    }
+  in
+  let substitued_toplevel_binding =
+    Msource.substitute toplevel_item_span
+      (`Logical (Lexing.split_pos subst_loc.loc_start))
+      (`Logical (Lexing.split_pos subst_loc.loc_end))
+      fresh_call
+    |> Msource.text
+  in
+  let untyped_expr = Untypeast.untype_expression expr in
+  let content =
+    match gen_binding_kind with
+    | Non_recursive ->
+      let fresh_let_binding =
+        generated_binding ~name:val_name ~body:untyped_expr
+        |> Format.asprintf "%a" Pprintast.structure_item
+      in
+      fresh_let_binding ^ "\n" ^ substitued_toplevel_binding
+    | Rec_and ->
+      let fresh_let_binding =
+        generated_binding ~name:val_name ~body:untyped_expr
+        |> Format.asprintf "%a" Pprintast.structure_item
+      in
+      let fresh_and_binding =
+        "and" ^ String.drop 3 fresh_let_binding (* Sorry *)
+      in
+      substitued_toplevel_binding ^ "\n" ^ fresh_and_binding
+  in
+  let selection_range =
+    let lnum =
+      match gen_binding_kind with
+      | Non_recursive -> toplevel_item.loc.loc_start.pos_lnum
+      | Rec_and -> toplevel_item.loc.loc_end.pos_lnum + String.length "\n"
+    in
+    let prefix_length =
+      match gen_binding_kind with
+      | Non_recursive ->
+        if is_recursive toplevel_item then String.length "let rec "
+        else String.length "let "
+      | Rec_and -> String.length "and "
+    in
+    { Location.loc_start = Lexing.make_pos (lnum, prefix_length);
+      loc_end = Lexing.make_pos (lnum, prefix_length + String.length val_name);
+      loc_ghost = false
+    }
+  in
+  { Query_protocol.loc = toplevel_item.loc; content; selection_range }
+
+let extract_const_to_toplevel ?extract_name expr ~expr_env ~toplevel_item =
+  let name =
+    Option.fold extract_name
+      ~none:(Default { basename = "const_name" })
+      ~some:(fun name -> Fixed name)
+  in
+  extract_to_toplevel
+    { expr;
+      expr_env;
+      toplevel_item;
+      name;
+      gen_binding_kind = Non_recursive;
+      generated_binding = Gen.toplevel_let;
+      generated_call = Gen.ident
+    }
+
+let extract_expr_to_toplevel ?extract_name expr ~expr_env ~toplevel_item
+    ~local_defs ~mconfig =
+  let is_function = function
+    | { Typedtree.exp_desc = Texp_function _; _ } -> true
+    | _ -> false
+  in
+  let { bounded_vars; gen_binding_kind } =
+    analyze_expr expr expr_env ~toplevel_item ~local_defs ~mconfig
+  in
+  let generated_binding, generated_call =
+    match bounded_vars with
+    | [] when Fun.negate is_function expr ->
```

Comment:
`not (is_function expr)`

Reply:
Fixed (`Fun.negate` is so cool 😩)

Reply:
For my point of view, if make more sense where it is used in partial application context. Like: `List.filter (Fun.negate predicate) list` (instead of `List.filter (fun x -> not (predicate x)) list`.